### DISCUSSION TITLE
can't restore cause system.replicas entries

### DISCUSSION THREAD

**Discussion by karthik-thiyagarajan** [Altinity/clickhouse-backup#1022] [Q&A] [answered]
I'm restoring the schema in each and every replica and shard (3 shard 3 replica - total 9 pods) and restoring the data in 0th replica of every shard. I'm trying to do only one table restoration and it fails with the below error. 

I checked the clickhouse-backup pod logs and i see the same error as given below. 

Any help would be appreciated. 

Restoration failure reason - 2024-10-08 08:08:21        2024-10-08 08:08:26     error   one of restoreDataRegular go-routine return error: can\'t attach data parts for table \'db.stats_local\': db.ap_stats_local can\'t restore cause system.replicas entries already exists and replication in progress from another replica, log_pointer=14648, log_max_index=14647, absolute_delay=0, queue_size=0
2024-10-08 08:18:24     2024-10-08 08:18:29     error   one of restoreDataRegular go-routine return error: can\'t attach data parts for table \'db.stats_local\': db.stats_local can\'t restore cause system.replicas entries already exists and replication in progress from another replica, log_pointer=14648, log_max_index=14647, absolute_delay=0, queue_size=0
jenkins@ip-10-58-130-135:/home/ubuntu$

**Comment by Slach** [comment, marked as answer]
> 'db.stats_local': db.stats_local can't restore cause system.replicas entries already exists and replication in progress from another replica, log_pointer=14648, log_max_index=14647, absolute_delay=0, queue_size=0

It means other replicas continue to write

Look to https://github.com/Altinity/clickhouse-backup/blob/master/Examples.md#restore

for one tables you need to execute in all replicas
`restore_remote --schema --table=db.table backup-name`
and then
run
`restore_remote --data --table=db.table backup-name` in first replica in each shard

**Reply by gavishdewangan** [reply]
that's what we are doing, restoring schema for all replicas and restoring data for first replica of each shard, still this error is given

**Reply by Slach** [reply]
Do you continue to insert into table during restore?

**Reply by karthik-thiyagarajan** [reply]
We stopped all writers.

**Reply by Slach** [reply]
Do you execute restoring process in parallel or use `kind: Job` manifest from https://github.com/Altinity/clickhouse-backup/blob/master/Examples.md#how-to-use-clickhouse-backup-in-kubernetes?

**Reply by karthik-thiyagarajan** [reply]
i stopped all writers (not readers) and the restored the schema  and data.  We used clickhouse:24.6 version with clickhouse-backup:2.5.2 version. Since we got the above issue, we even upgraded to clickhouse-backup:2.6.1 as well but with no luck. 

Here is the pseudo code and we are using a k8s job to restore the content for a specific table. 

kind: Job 

    if [[ "ALL" == "${TABLE_NAMES}" ]]; then 
        echo "All tables from all databases will be restored..."; 
        TABLE_NAME_PATTERN=" "; 
    else
        echo "Specific tables only will be restored - ${TABLE_NAMES}";
        TABLE_NAME_PATTERN=" --table ${TABLE_NAMES}";
    fi;

    for every shard and every replica : 
        QUERY="INSERT INTO system.backup_actions(command) VALUES('restore_remote --rm --schema ${TABLE_NAME_PATTERN} ${BACKUP_NAMES[$SERVER]}')";
        STATUS=$(echo "${QUERY}" | curl -s -H "X-ClickHouse-User:${BACKUP_USER}" -H "X-ClickHouse-Key:${BACKUP_PASSWORD}" http://"${SERVER}:${CLICKHOUSE_PORT}"/ --data-binary @-); 

    for every shard and every replica : 
        QUERY="INSERT INTO system.backup_actions(command) VALUES('restore_remote delete local ${TABLE_NAME_PATTERN} ${BACKUP_NAMES[$SERVER]}')";
        STATUS=$(echo "${QUERY}" | curl -s -H "X-ClickHouse-User:${BACKUP_USER}" -H "X-ClickHouse-Key:${BACKUP_PASSWORD}" http://"${SERVER}:${CLICKHOUSE_PORT}"/ --data-binary @-); 

    for every 0th replica of every shard: 
        QUERY="INSERT INTO system.backup_actions(command) VALUES('restore_remote --data ${TABLE_NAME_PATTERN} ${BACKUP_NAMES[$SERVER]}')"; 
        STATUS=$(echo "${QUERY}" | curl -s -H "X-ClickHouse-User:${BACKUP_USER}" -H "X-ClickHouse-Key:${BACKUP_PASSWORD}" http://"${SERVER}:${CLICKHOUSE_PORT}"/ --data-binary @-);

**Reply by karthik-thiyagarajan** [reply]
All the 0th replica of every shard will be initiating the restoration process in parallel.

**Reply by Slach** [reply]
> ```
> QUERY="INSERT INTO system.backup_actions(command) VALUES('restore_remote delete local ${TABLE_NAME_PATTERN} ${BACKUP_NAMES[$SERVER]}')";
> ```

`restore_remote delete local` or just `delete local`?

As a workaround you could try
```
clickhouse:
  check_replicas_before_attach: false 
```

but need to check about duplicates after restore

`restore --schema`
shall execute DROP TABLE .. SYNC and clean `system.replication_queue`, `system.replicas` and zookeeper data and set log_pointer as =0

check
```sql
SELECT * FROM clusterAllReplicas('your-cluster-name',system.backup_actions)
```
to ensure you didn't execute restore --data on all replicas

could you share?
```sql
SELECT engine_full FROM system.tables WHERE database='db' AND table='stats_local'
```

need ensure do you use {shard} macros in replication_path parameter

**Reply by gavishdewangan** [reply]
1) it is delete local
2) this workaround worked but we are facing another error now
Restoration failure reason - 2024-10-10 04:07:41	2024-10-10 04:07:47	error	one of restoreDataRegular go-routine return error: can\'t attach data parts for table \'db.stats_local\': code: 242, message: Table is in readonly mode (replica path: /clickhouse/tables/cnxcluster-01/0/db/stats/replicas/chi-clickhouse-server-cnxcluster-01-0-0)

3) We are executing --data for first replica only

4)  ReplicatedMergeTree('/clickhouse/tables/{cluster}/{shard}/db/stats', '{replica}')

**Reply by Slach** [reply]
Looks like you doing something wrong outside `clickhouse-backup`
check `grep -C 50 -E "read.{0,1}only" /var/log/clickhouse-server/clickhouse-server.log`

**Reply by Slach** [reply]
could share full logs from `clickhouse-backup` container, all replicas in one shard for restore command?

**Reply by karthik-thiyagarajan** [reply]
We dropped the table completely and then added the env flag "check_replicas_before_attach" while doing the restore and it seems to be working now. Can you explain what does this flag do and why was it not working previously and now it s working ? We have stopped all the writers definitely and I'm pretty sure this restore was working some 6 months ago though not sure what version we were using at that time but this seems to be happening from 2.5.20 and above onwards including the current version 2.6.1

**Reply by Slach** [reply]
This flag check `system.replicas` state and if it show us "replication in progress" return error 

during `restore --schema`
in all replicas clickhouse-backup should execute `DROP TABLE ... SYNC` in all replicas
add `log_level: debug` in your backup config and check it in `clickhouse-backup` log
